### PR TITLE
Provision to export inventory

### PR DESCRIPTION
**Pull request by tarunbhardwaj**

### CODE REVIEW

File: channel.py [hunk 1]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+        """
+        ChannelListing = Pool().get('product.product.channel_listing')
+        Location = Pool().get('stock.location')
+
+        domain = [
+            ('channel', '=', self),
+            ('state', '=', 'active')
+        ]
+        product_ids = set([])
+        locations = map(int, Location.search([('type', '=', 'storage')]))
```
Comment (sharoonthomas) [inline channel.py]: I dint think you should filter by locations at all.

File: channel.py [hunk 2]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+
+        if self.last_inventory_export_time:
+            # For optimization this sql query is to find products
+            # that are updated after last inventory export
+            Transaction().cursor.execute(
+                "select distinct(product) from stock_move where "
+                "coalesce(write_date, create_date) > %s and "
+                "(from_location = ANY(%s) or "
```
Comment (sharoonthomas) [inline channel.py]: remove the conditions by location

File: channel.py [hunk 3]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+                "to_location = ANY(%s)) "
+                "and state = 'done'",
```
Comment (sharoonthomas) [inline channel.py]: even draft applies if the channel looks at forecasted inventory to decide on the availability.

File: channel.py [hunk 4]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+            ('state', '=', 'active')
+        ]
+        product_ids = set([])
+        locations = map(int, Location.search([('type', '=', 'storage')]))
+
+        if self.last_inventory_export_time:
+            # For optimization this sql query is to find products
+            # that are updated after last inventory export
+            Transaction().cursor.execute(
+                "select distinct(product) from stock_move where "
```
Comment (sharoonthomas) [inline channel.py]: instead of distinct, use group by

File: channel.py [hunk 5]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+        Since external channels are implemented by downstream modules, it is
+        the responsibility of those channels to implement exporting or call
+        super to delegate.
+        """
+        # TODO: check if inventory export is allowed for this channel
+        for listing in self.get_listings_to_export_inventory():
+            listing.export_inventory()
+
+    @classmethod
+    def export_inventory_using_cron(cls):  # pragma: nocover
```
Comment (sharoonthomas) [inline channel.py]: export_inventory_from_cron

File: channel.py [hunk 6]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+        Location = Pool().get('stock.location')
+
+        domain = [
+            ('channel', '=', self),
+            ('state', '=', 'active')
+        ]
+        product_ids = set([])
+        locations = map(int, Location.search([('type', '=', 'storage')]))
+
+        if self.last_inventory_export_time:
```
Comment (sharoonthomas) [inline channel.py]: else ?

File: channel.py [hunk 7]
```diff
@@ -303,6 +308,85 @@ def export_product_catalog(self):
...
+                ], [
+                    ('create_date', '!=', None),
+                    ('create_date', '>=', self.last_order_import_time),
+                ]]
+            )
+            domain.append(extra_domain)
+
+        # XXX: Optimize me (may be one sql query to return listing will do)
+        return ChannelListing.search(domain)
+
```
Comment (sharoonthomas) [inline channel.py]: add a class method `export_inventory_bulk` which can then call individual lines or handle records in bulk

File: channel.py (hunk 6 above)
Comment (tarunbhardwaj) [inline channel.py]: else no filter on products and time and it will pickup all products

File: channel.py (hunk 7 above)
Comment (tarunbhardwaj) [inline channel.py]: added

**Comment by prakashpp** [comment]
LGTM

**Pull request merged**